README
== Description
    Just another HTML table generator for ruby.
    Generate HTML tables with ease (HTML4, XHTML and HTML5).
    Generate portrait, landscape and other rotated views. Handles
    rotating table attributes via HTML-AutoTag.

== Installation
    gem install Spreadsheet-HTML

== Synopsis
    require 'Spreadsheet/HTML'

    data = Array[ [1,2,3], [4,5,6], [7,8,9] ]
    generator = Spreadsheet::HTML.new()
    puts generator.generate( data )
    puts generator.generate( 'data' => data )

    generator = Spreadsheet::HTML.new( 'data' => data, 'indent' => "\t" )
    puts generator.portrait( 'encodes' => 1 )
    puts generator.landscape( 'encode' => 1 )

    puts generator.generate( 'tgroups' => 1 )
    puts generator.generate( 'tgroups' => 2, 'indent' => nil )

    puts generator.generate( 'tr' => { 'class' => %w{ odd even } } )

    # See t/ directory for more examples

== Dependencies
    HTML-AutoTag

== Methods
    With the exception of new, all methods return an HTML table as a string.

new( params )
    Constructs a new generator configured with given params. These params will be used
    for each call to a generator method. Any params specified in the constructor
    may be overriden in a subsequent call to a generator method (which are listed next).

generate( params )
    Generates an HTML table with headings positioned at the top.

portrait( params )
    Generates an HTML table with headings positioned at the top.

north( params )
    Generates an HTML table with headings positioned at the top.

landscape( params )
    Generates an HTML table with headings positioned at the left.
    
west( params )
    Generates an HTML table with headings positioned at the left.
    
south( params )
    Generates an HTML table with headings positioned at the bottom.

east( params )
    Generates an HTML table with headings positioned at the right.

== Literal Parameters
    Literal Parameters provides the means to modify the macro aspects of the table,
    such as indentation, encoding, data source and table orientation.

data
    The data to be rendered as an HTML table. Array of Arrays.

  'data' = > [ [ 1, 2, 3 ], [ 4, 5, 6 ], [7, 8, 9 ] ]

tgroups
    Integer (0, 1 or 2). Group table rows into <thead>, <tbody> and <tfoot> sections.
    When tgroups is set to 1, the <tfoot> section is omitted. The last row of the data
    is found at the end of the <tbody> section instead. (loose) When tgroups is set to 2,
    the <tfoot> section is found in between the <thead> and <tbody> sections. (strict)

  'tgroups' => 1
  'tgroups' => 2

indent
    The string to indent each row by. Defaults to undefined which produces no indentation.
    Automatically adds newlines when set to any defined value.

  'indent' => '    '
  'indent' => "\t"

level
    Positive integer. The level to start indentation at. This is useful for matching the nesting styles of
    original HTML text that you may wish in insert into. (A value of 4 says "apply the repitition operator
    to the value of indent 4 times.)

  'level' => 3

empty
    String. Render any empty cells with this value. Defaults to &nbps; 

  'empty' => ''

encode
    Boolean. Whether or not to Encode entities.

  'encode' => 1

encodes
    String. Encode these HTML entities. Pass string with chars you want encoded
    or leave blank for default control and high bit chars and <>!'"

  'encodes' => '<>'

matrix
    Boolean. Render the headings row with only <td> tags, no <th> tags.

  'matrix' => 1

headless
    Boolean. Render the table without the headings row at all.

  'headless' => 1

theta
    Rotates table clockwise for positive values and counter-clockwise for negative values.
    Default is 0: headings at top. 90 yields headings at right, 180 yields headings at bottom.
    270 yields headings at left. To achieve landscape, use -270 (or use the landscape() method).

  'theta' => -270

flip
    Flips the table horizontally from the perspective of the headings "row" by
    negating the value of theta.

  'flip' => 1

pinhead
    Works in conjunction with theta to ensure reporting readability. Without it, south() and east()
    would have data cells arranged in reverse order.

  'pinhead' => 1

fill
    Can be used instead of data to create a "blank" table or with data to pad rows and cells.

  'fill' => '8x12' # 8 rows, 12 columns

== Dynamic Parameters
    Dynamic parameters provide a means to control the micro elements of the table,
    such as modifying headings by their name and rows and columns by their indices.
    They contain leading underscores to seperate them from literal and tag parameters.
    They accepts Hashes for use as tag attributes, Procs or lambdas for altering the
    CDATA of the cell, or Arrays that contain either (just like the 'td' and 'th' Tag
    Parameters described below).

_rX
    Apply these attributes to row X (zero index based).

  '_r2' => { 'class' => 'third-row' }
  '_r2' => lambda { |c| c.upcase }
  '_r2' => [ lambda { |c| c.upcase }, { 'class' => 'third-row' } ]

_cX
    Apply these attributes to colum X (zero index based). You can also alias any
    column by the value of the heading name in that column prepended with underscore (_)

  '_c4' => { 'class' => 'fifth-column' }
  '_c4' => lambda { |c| c.upcase }
  '_c4' => [ lambda { |c| c.upcase }, { 'class' => 'fifth-column' } ]

  '_occupation' => { 'class' => 'foo' }
  '_salary' => lambda { |c| c.capitalize }

_rXcY
    Apply these attributes to colum Y in row X (zero index based).

  '_r2c4' => { 'class' => 'third-row-fifth-column' }
  '_r2c4' => lambda { |c| c.upcase }
  '_r2c4' => [ lambda { |c| c.upcase }, { 'class' => 'third-row-fifth-column' } ]

== Tag Parameters
    Tag Parameters provide a means to control the attributes of the table's tags, and in
    the case of <th> and <td> the contents via callback subroutines. Although
    similar in form, they are differentiated from litertal parameters because they share
    the names of the actual HTML table tags.

table
    Hash. Apply these attributes.

  'table' => { 'class' => 'spreadsheet' }

thead
    Hash. Apply these attributes.

  'thead' => { 'class' => 'headings' }

tfoot
    Hash. Apply these attributes.

  'tfoot' => { 'class' => 'footings' }

tbody
    Hash. Apply these attributes.

  'tbody' => { 'class' => 'bodyings' }

tr
    Hash. Apply these attributes.

  'tr => { 'class' => %w{ odd even } }

th
    <th> and <td> are the only Tag Parameters that also accept callback methods.

  'th' => { 'style' => { 'color' => %w{ red green blue } } }
  'th' => lambda { |c| c.upcase }
  'th' => [ { 'style' => { 'color' => %w{ red green blue } }, lambda { |c| c.upcase } ]

td
    <th> and <td> are the only Tag Parameters that also accept callback methods.

  'td' => { 'style' => { 'color' => %w{ red green blue } } }
  'td' => lambda { |c| c.upcase }
  'td' => [ { 'style' => { 'color' => %w{ red green blue } }, lambda { |c| c.upcase } ]

caption
    Caption is special in that you can either pass a string to be used as CDATA or a hash whose only
    key is the string to be used as CDATA.

  'caption' => 'Just Another Title'

  'caption' => { 'A Title With Attributes' => { 'align' => 'bottom' } }

colgroup
    Add colgroup tag(s) to the table. Use an AoH for multiple.

  colgroup => { 'span => '2', 'style' => { 'background-color' => 'orange' } }

  colgroup => Array[ { 'span => '20' }, { 'span' => '1', 'class' => 'end' } ]

col
    Add col tag(s) to the table. Use an AoH for multiple. Wraps tags within a colgroup tag. Same usage as colgroup.

  col => { 'span' => '2', 'style' => { 'background-color' => 'orange' } }

  col => Array[ { 'span' => 20 }, { 'span' => '1', 'class' => 'end' } ]

thead.tr
    When tgroups is 1 or 2, this tag parameter is available to control the attributes of
    the <tr> tag within the <thead> group.

  'thead.tr' => { 'class' => 'body-cell' }

tfoot.tr
    When tgroups is 2, this tag parameter is available to control the attributes of
    the <tr> tag within the <tfoot> group.

  'tbody.tr' => { 'class' => 'body-cell' }

== License
    MIT

== Warranty
   This package is provided "as is" and without any express or
   implied warranties, including, without limitation, the implied
   warranties of merchantability and fitness for a particular purpose.

== Author
    Jeff Anderson
    jeffa@cpan.org


Changes
Revision history for html-autotag-ruby

1.0.4

1.0.3
    - require HTML::AutoTag v1.0.11
    - update copyright date

1.0.1
    - require HTML::AutoTag v1.0.9
    - update copyright date

1.0.0
    - production ready
    - corrected docs
    - require HTML::AutoTag v1.0.7

0.0.11
    - added col and colgroup params
    - added caption param

0.0.10
    - implemented lambda/proc for dynamic and tag params

0.0.9
    - implemented Dynamic Tags by headings
    - handle invalid fill params more gracefully

0.0.8
    - implemented fill param
    - "defactored" HTML Encoding into this module
    - refactored out default {} for attrs (HTML::AutoAttr handles them)
    - require HTML::AutoTag v1.0.6

0.0.7
    - fix Ruby dependecy

0.0.6
    - implemented empty param
    - implemented encodes param
    - implemented row padding and truncation

0.0.5
    - moved docs to README (trying to render for rubydoc.info site)
    - just a note -- having a very hard time learning how to document ruby

0.0.4
    - added docs location to gemspec

0.0.3
    - added docs (hopefully in the right place)

0.0.2
    - no real changes, just ensuring that all new files are published

0.0.1
    - First version, released on an unsuspecting world.


t/02-run.rb
require "test/unit"
require "Spreadsheet/HTML"

class Test_Run < Test::Unit::TestCase

  def test_instance

    data = Array[[1,'a'], [2,'b']]
    html = '<table><tr><th>1</th><th>a</th></tr><tr><td>2</td><td>b</td></tr></table>'

    assert_equal(
        html,
        Spreadsheet::HTML.new( 'data' => data ).generate(),
        "via constructor"
    )

    gen = Spreadsheet::HTML.new

    assert_equal(
        html,
        gen.generate( [1,'a'], [2,'b'] ),
        "two array refs"
    )

    assert_equal(
        html,
        gen.generate( data ),
        "one array ref"
    )

    assert_equal(
        html,
        gen.generate( 'data' => data ),
        "one named arg"
    )

  end

  def test_class

    data = Array[[1,'a'], [2,'b']]
    html = '<table><tr><th>1</th><th>a</th></tr><tr><td>2</td><td>b</td></tr></table>'

    assert_equal(
        html,
        Spreadsheet::HTML.gen( [1,'a'], [2,'b'] ),
        "two array refs"
    )

    assert_equal(
        html,
        Spreadsheet::HTML.gen( data ),
        "one array ref"
    )

    assert_equal(
        html,
        Spreadsheet::HTML.gen( 'data' => data ),
        "one named arg"
    )

  end

end


t/01-load.rb
require "test/unit"

class TestLoad < Test::Unit::TestCase

  def test_load
    assert_nothing_raised { require "Spreadsheet/HTML.rb" }
  end

end


lib/HTML/Spreadsheet.rb
require "HTML/Spreadsheet/version"
require "HTML/AutoTag"

module HTML
    class Spreadsheet

        def self.gen( *args )
            self.new.generate( *args )
        end

        def generate( *args )
            params = _process( args )

            if params['theta'] and params['flip'] 
                params['theta'] *= -1 
            end

            if !params['theta']           # north
                if params['flip']
                    params['data'] = params['data'].map {|a| a.reverse }
                end

            elsif params['theta'] == 90   # east
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].transpose.map{|a| a.push( a.shift ) }
                else
                    params['data'] = params['data'].transpose.map{|a| a.reverse }
                end

            elsif params['theta'] == -90
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].transpose.reverse.map {|a| a.push( a.shift ) } 
                else
                    params['data'] = params['data'].transpose.reverse.map {|a| a.reverse }
                end

            elsif params['theta'] == 180
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].push( params['data'].shift ).map {|a| a.reverse }
                else
                    params['data'] = params['data'].reverse.map {|a| a.reverse }
                end

            elsif params['theta'] == -180 # south
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].push( params['data'].shift )
                else
                    params['data'] = params['data'].reverse
                end

            elsif params['theta'] == 270
                params['data'] = params['data'].transpose.reverse

            elsif params['theta'] == -270 # west
                params['data'] = params['data'].transpose

            end

            return _make_table( params )
        end

        def initialize( *args )
            (args[0] || []).each do |key,val|
                self.instance_eval { class << self; self end }.send(:attr_accessor, key)
                self.send( "#{key}=", val )
            end
        end

        def _make_table( params )
            cdata = [] # insert caption and colgroup
            cdata.push( params['data'].map { |c| { 'tag' => 'tr', 'attr' => {}, 'cdata' => c } } )
            return params['auto'].tag( 'tag' => 'table', 'attr' => {}, 'cdata' => cdata )
        end

        def _process( args )
            params = _args( args )

            params['data'].shift if params['headless']

            data = []
            tag  = ( params['matrix'] or params['headless'] ) ? 'td' : 'th'

            params['data'].each do |row|
                r = []
                row.each do |col|
                    r.push( { 'tag' => tag, 'cdata' => col.to_s } )
                end
                data.push( r )
                tag = 'td'
            end
            params['data'] = data

            return params
        end

        def _args( things )

            data   = []
            args   = []
            while !things.empty?

                if things[0].kind_of?(Array)
                    data.push( things.shift() )
                    if things[0].kind_of?(Array)
                        data.push( things.shift() )
                    else
                        args.push( things.shift() ) if !things.empty?
                        args.push( things.shift() ) if !things.empty?
                    end
                else
                    args.push( things.shift() ) if !things.empty?
                    args.push( things.shift() ) if !things.empty?
                end

            end

            params = {}
            (args[0] || []).each do |key,val|
                params[key] = val
            end

            self.instance_variables.each do |attr|
                params[attr[1..-1]] = self.instance_variable_get attr
            end

            params['auto'] = HTML::AutoTag.new(
                'encodes'   => params['encodes'],
                'indent'    => params['indent'],
                'level '    => params['level'],
                'sorted'    => params['attr_sort'],
            )

            if !params['data'] and data[0].kind_of?(Array)
                data = [ data ] if !data[0][0].kind_of?(Array)
                params['data'] = data[0]
            end

            params['data'] = params['data'].clone

            return params
        end

    end

end


lib/Spreadsheet/HTML.rb
require "Spreadsheet/HTML/version"
require "HTML/AutoTag"

# i cannot instantiate any HTML::foo inside any bar::HTML?!?
module Auto
    class Tag < HTML::AutoTag
    end
end
module Enco
    class Der < HTML::Encoder
    end
end

module Spreadsheet
    class HTML

        def self.gen  ( *args ) self.new.generate( *args ) end

        def portrait  ( *args ) generate( *args, 'theta' => 0 ) end
        def landscape ( *args ) generate( *args, 'theta' => -270, 'tgroups' => 0 ) end

        def north     ( *args ) generate( *args, 'theta' => 0 ) end
        def west      ( *args ) generate( *args, 'theta' => -270, 'tgroups' => 0 ) end
        def east      ( *args ) generate( *args, 'theta' =>   90, 'tgroups' => 0, 'pinhead' => 1 ) end
        def south     ( *args ) generate( *args, 'theta' => -180, 'tgroups' => 0, 'pinhead' => 1 ) end

        def generate( *args )
            params = _process( args )

            if params['theta'] and params['flip'] 
                params['theta'] *= -1 
            end

            if !params['theta']           # north
                if params['flip']
                    params['data'] = params['data'].map {|a| a.reverse }
                end

            elsif params['theta'] == 90   # east
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].transpose.map{|a| a.push( a.shift ) }
                else
                    params['data'] = params['data'].transpose.map{|a| a.reverse }
                end

            elsif params['theta'] == -90
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].transpose.reverse.map {|a| a.push( a.shift ) } 
                else
                    params['data'] = params['data'].transpose.reverse.map {|a| a.reverse }
                end

            elsif params['theta'] == 180
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].push( params['data'].shift ).map {|a| a.reverse }
                else
                    params['data'] = params['data'].reverse.map {|a| a.reverse }
                end

            elsif params['theta'] == -180 # south
                if params['pinhead'] and !params['headless']
                    params['data'] = params['data'].push( params['data'].shift )
                else
                    params['data'] = params['data'].reverse
                end

            elsif params['theta'] == 270
                params['data'] = params['data'].transpose.reverse

            elsif params['theta'] == -270 # west
                params['data'] = params['data'].transpose

            end

            return _make_table( params )
        end

        def initialize( *args )
            (args[0] || []).each do |key,val|
                self.instance_eval { class << self; self end }.send(:attr_accessor, key)
                self.send( "#{key}=", val )
            end
        end

        def _make_table( params )
            cdata = Array[]

            if params.has_key?( 'caption' )
                cdata.push( _tag( 'caption', params['caption'] ) )
            end

            if params.has_key?( 'colgroup' ) or params.has_key?( 'col' )
                cdata.concat( _colgroup( params ) )
            end

            if params['tgroups'] && params['tgroups'] > 0

                body = params['data']
                # TODO: is this a bug? data vs params['data']
                head = body.shift() unless params['matrix'] && data.size() > 2
                foot = body.pop() if !params['matrix'] && params['tgroups'] > 1 and data.size() > 2

                head_row  = { 'tag' => 'tr', 'attr' => params['thead.tr'], 'cdata' => head }
                foot_row  = { 'tag' => 'tr', 'attr' => params['tfoot.tr'], 'cdata' => foot }
                body_rows = body.map{ |r| { 'tag' => 'tr', 'attr' => params['tr'], 'cdata' => r } }

                cdata.push({ 'tag' => 'thead', 'attr' => params['thead'], 'cdata' => head_row }) if head
                cdata.push({ 'tag' => 'tfoot', 'attr' => params['tfoot'], 'cdata' => foot_row }) if foot
                cdata.push({ 'tag' => 'tbody', 'attr' => params['tbody'], 'cdata' => body_rows })

            else
                cdata.concat( params['data'].map { |c| { 'tag' => 'tr', 'attr' => params['tr'], 'cdata' => c } } )
            end

            return params['auto'].tag( 'tag' => 'table', 'attr' => params['table'], 'cdata' => cdata )
        end

        def _process( args )
            params = _args( args )

            # headings is an alias for _r0
            params['_r0'] = params['headings']

            index = {}
            if params['data'][0].size()
                for i in 0 .. params['data'][0].size() - 1
                    key = params['data'][0][i] || ''
                    index["_#{key}"] = i
                end

                params.keys.grep( /^_/ ) do |key|
                    k = index[key] 
                    params["_c#{k}"] = params[key] if index.has_key?( key )
                end
            end

            empty = params['empty'] || '&nbsp;'
            tag   = ( params['matrix'] or params['headless'] ) ? 'td' : 'th'

            encoder = Enco::Der.new
            for r in 0 .. params['_max_rows'] - 1

                unless params['_layout']
                    params['data'][r] ||= []
                    (params['_max_cols'] - params['data'][r].size).times { params['data'][r].push( nil ) }  # pad
                    (params['data'][r].size - params['_max_cols']).times { params['data'][r].pop }          # truncate
                end

                row = []
                for c in 0 .. params['_max_cols'] - 1

                    attr  = params[tag] || {}
                    cdata = params['data'][r][c].to_s

                    [ tag, "_c#{c}", "_r#{r}", "_r#{r}c#{c}" ].each do |dyna_param|
                        if params.has_key?( dyna_param )
                            ( cdata, attr ) = _extrapolate( cdata, attr, params[dyna_param] );
                        end
                    end 

                    cdata = encoder.encode( cdata, params['encodes'] ) if params['encode'] or !params['encodes'].to_s.empty?
                    cdata = cdata.gsub( /^\s*$/, empty.to_s )
                    row.push( { 'tag' => tag, 'attr' => attr, 'cdata' => cdata } )

                end

                params['data'][r] = row
                tag = 'td'
            end

            params['data'].shift if params['headless']

            return params
        end

        def _args( things )

            data   = []
            args   = []
            while !things.empty?

                if things[0].kind_of?(Array)
                    data.push( things.shift() )
                    if things[0].kind_of?(Array)
                        data.push( things.shift() )
                    else
                        args.push( things.shift() ) if !things.empty?
                        args.push( things.shift() ) if !things.empty?
                    end
                else
                    args.push( things.shift() ) if !things.empty?
                    args.push( things.shift() ) if !things.empty?
                end

            end

            params = {}
            self.instance_variables.each do |attr|
                params[attr[1..-1]] = self.instance_variable_get attr
            end

            (args || []).each do |hash|
                hash.each do |key,val|
                    params[key] = val
                end
            end

            params['auto'] = Auto::Tag.new(
                'indent'    => params['indent'],
                'level'     => params['level'],
                'sorted'    => params['attr_sort'],
            )

            if !params['data'] and data[0].kind_of?(Array)
                data = [ data ] if !data[0][0].kind_of?(Array)
                params['data'] = data[0]
            end

            params['data'] = [ [ nil ] ] if params['data'].nil?

            params['_max_rows'] = params['data'].size    || 1
            params['_max_cols'] = params['data'][0].size || 1
            params['data']      = params['data'].clone

            if params['fill']
                (row,col) = params['fill'].split(/\D/)    
                if row.to_i > 0 && col.to_i > 0
                    params['_max_rows'] = row.to_i if row.to_i > params['_max_rows']
                    params['_max_cols'] = col.to_i if col.to_i > params['_max_cols']
                end
            end

            return params
        end

        def _extrapolate( cdata, orig_attr, thingy )
            new_attr = {}
            thingy = [ thingy ] unless thingy.kind_of?(Array)
            thingy.each do |t|
                if t.kind_of?(Hash)
                    new_attr = t
                elsif t.kind_of?(Proc)
                    cdata = t.call( cdata )
                end
            end

            attr = {}
            orig_attr.each { |key,val| attr[key] = val } if orig_attr.kind_of?(Hash)
            new_attr.each  { |key,val| attr[key] = val }

            return [ cdata, attr ]
        end

        def _colgroup( params )
            colgroup = Array[]
            params['col'] = Array[ params['col'] ] if params['col'].kind_of?(Hash)

            if params['col'].kind_of?(Array)
                if params['colgroup'].kind_of?(Array)
                    colgroup = params['colgroup'].map{ |cg| 
                        {
                            'tag'   => 'colgroup',
                            'attr'  => cg,
                            'cdata' => params['col'].map{ |attr| { 'tag' => 'col', 'attr' => attr } }
                        }
                    }
                else
                    colgroup = {
                        'tag'   => 'colgroup',
                        'attr'  => params['colgroup'],
                        'cdata' => params['col'].map{ |attr| { 'tag' => 'col', 'attr' => attr } }
                    }
                end
            else
                params['colgroup'] = Array[ params['colgroup'] ] if params['colgroup'].kind_of?(Hash)
                if params['colgroup'].kind_of?(Array)
                    colgroup = params['colgroup'].map{ |attr| { 'tag' => 'colgroup', 'attr' => attr } }
                end
            end

            colgroup = Array[colgroup] if colgroup.kind_of?(Hash)
            return colgroup
        end

        def _tag( tag, cdata )
            tag = { 'tag' => tag, 'cdata' => cdata }
            if cdata.kind_of?( Hash )
                tag['cdata'] = cdata.keys[0]
                tag['attr']  = cdata.values[0]
            end
            return tag
        end

    end

end

=begin rdoc

== AUTHOR

Jeff Anderson, <jeffa at cpan.org>

== LICENSE AND COPYRIGHT

Copyright 2017 Jeff Anderson. (See License.md shipped with distro)

=end
